File: python/periph/chips/motor/drv8830.py
"""DRV8830 — Low-voltage motor driver with I2C interface (Texas Instruments).

H-bridge driver for a single brushed DC motor, controlled entirely over I2C.
Instead of a duty cycle, the host commands a target output voltage; the chip
PWM-regulates the bridge to hold that average voltage across the winding
regardless of supply sag. Nine selectable addresses (0x60-0x68) via the
tri-state A0/A1 strap pins. The same driver file is used on MicroPython,
CircuitPython, and Linux hosts.

Args:
    connection: Configured I2C connection pointing at the device
        (0x60-0x68, per the board's A0/A1 strapping).
"""

try:
    import threading as _threading
    _LINUX = True
except ImportError:
    _LINUX = False


I2C_ADDRESS = 0x60

# Internal reference voltage, typical (datasheet: 1.235-1.335 V).
VREF = 1.285

# Valid VSET DAC codes (0x00-0x05 are reserved).
VSET_MIN = 6
VSET_MAX = 63

# Direction names returned by read_output().
DIRECTION_COAST = 'coast'
DIRECTION_REVERSE = 'reverse'
DIRECTION_FORWARD = 'forward'
DIRECTION_BRAKE = 'brake'

_DIRECTIONS = (DIRECTION_COAST, DIRECTION_FORWARD, DIRECTION_REVERSE, DIRECTION_BRAKE)


def _voltage_to_vset(voltage):
    """Map |voltage| to a VSET code; 0 means coast (below the vset=6 floor)."""
    vset = int(abs(voltage) * 16.0 / VREF + 0.5)
    if vset < VSET_MIN:
        return 0
    if vset > VSET_MAX:
        return VSET_MAX
    return vset


def _vset_to_voltage(vset):
    if vset < VSET_MIN:
        return 0.0
    return VREF * vset / 16.0


class DRV8830Minimal:
    """DRV8830 low-voltage motor driver — minimal interface.

    Drives the motor at a regulated voltage in either direction; no
    configuration required beyond the connection.

    Args:
        connection: Configured I2C connection pointing at the device.
    """

    _REG_CONTROL = 0x00
    _REG_FAULT   = 0x01

    # CONTROL (0x00) bits.
    _CTRL_IN1 = 0x01
    _CTRL_IN2 = 0x02

    # FAULT (0x01) bits.
    _FAULT_FAULT  = 0x01
    _FAULT_OCP    = 0x02
    _FAULT_UVLO   = 0x04
    _FAULT_OTS    = 0x08
    _FAULT_ILIMIT = 0x10
    _FAULT_CLEAR  = 0x80

    def __init__(self, connection):
        self._connection = connection
        self._read_reg(self._REG_CONTROL)  # presence check; no writes needed

    def _write_reg(self, reg, value):
        self._connection.write(bytes([reg, value & 0xFF]))

    def _read_reg(self, reg):
        return self._connection.write_read(bytes([reg]), 1)[0]

    def drive(self, voltage):
        """Drive the motor at a regulated output voltage.

        Writes VSET and IN1/IN2 together in one CONTROL write. A magnitude
        below the chip's ~0.48 V floor is treated as 0 V (coast); above
        ~5.06 V it is clamped to the maximum code.

        Args:
            voltage: Signed target voltage in volts — positive = forward,
                negative = reverse, 0.0 = standby/coast.
        """
        vset = _voltage_to_vset(voltage)
        if vset == 0:
            self._write_reg(self._REG_CONTROL, 0x00)
        elif voltage > 0:
            self._write_reg(self._REG_CONTROL, (vset << 2) | self._CTRL_IN1)
        else:
            self._write_reg(self._REG_CONTROL, (vset << 2) | self._CTRL_IN2)

    def brake(self):
        """Short-brake the motor (IN1 = IN2 = 1, both outputs high)."""
        self._write_reg(self._REG_CONTROL, self._CTRL_IN1 | self._CTRL_IN2)

    def stop(self):
        """Put the bridge in standby/coast (IN1 = IN2 = 0) — same as drive(0.0)."""
        self._write_reg(self._REG_CONTROL, 0x00)


class DRV8830Full(DRV8830Minimal):
    """DRV8830 full interface — extends Minimal with raw CONTROL access,
    output read-back, fault reporting/clearing, and the FAULTn interrupt API.

    Faults are never cleared implicitly: a latched OCP/ILIMIT fault also
    disables the H-bridge, so clearing is always an explicit clear_fault().

    Args:
        connection: Configured I2C connection pointing at the device.
    """

    def __init__(self, connection):
        super().__init__(connection)
        self._callback = None
        self._int_pin_used = None
        self._poll_stop = False
        self._poll_thread = None

    def set_output(self, vset, in1, in2):
        """Write the CONTROL register from raw fields.

        Args:
            vset: VSET DAC code, 6-63 (0-5 are reserved).
            in1: H-bridge input 1.
            in2: H-bridge input 2.

        Raises:
            ValueError: if vset is outside 6-63.
        """
        if vset < VSET_MIN or vset > VSET_MAX:
            raise ValueError('vset must be 6-63')
        value = (vset << 2) | (self._CTRL_IN1 if in1 else 0) | (self._CTRL_IN2 if in2 else 0)
        self._write_reg(self._REG_CONTROL, value)

    def read_output(self):
        """Read back and decode the CONTROL register.

        Returns:
            tuple: (voltage, direction) — voltage is the commanded magnitude
            in volts (0.0 for a reserved/zero VSET code); direction is one of
            'forward', 'reverse', 'coast', 'brake'.
        """
        ctrl = self._read_reg(self._REG_CONTROL)
        return (_vset_to_voltage(ctrl >> 2), _DIRECTIONS[ctrl & 0x03])

    def read_fault(self):
        """Read the FAULT register without clearing it.

        Returns:
            tuple: (fault, ocp, uvlo, ots, ilimit), all bool.
        """
        f = self._read_reg(self._REG_FAULT)
        return (bool(f & self._FAULT_FAULT), bool(f & self._FAULT_OCP),
                bool(f & self._FAULT_UVLO), bool(f & self._FAULT_OTS),
                bool(f & self._FAULT_ILIMIT))

    def clear_fault(self):
        """Clear all fault status bits (CLEAR = 1).

        Re-enables the H-bridge if an OCP/ILIMIT fault had latched it off.
        """
        self._write_reg(self._REG_FAULT, self._FAULT_CLEAR)

    def on_interrupt(self, callback, int_pin=None):
        """Subscribe to fault interrupts (FAULTn, active-low).

        Delivery: if int_pin is given, it is wired directly. Otherwise falls
        back to connection.int_pin, or a 5 ms polling thread on Linux that
        fires on each new fault (FAULT bit rising). Does not clear the fault.

        Args:
            callback: Callable((fault, ocp, uvlo, ots, ilimit)) — same tuple
                as read_fault().
            int_pin: Optional InputPin for this call, overriding connection.int_pin.
        """
        self._callback = callback
        pin = int_pin if int_pin is not None else getattr(self._connection, 'int_pin', None)
        self._int_pin_used = pin
        if pin is not None:
            from periph.connection.input_pin import InputPin
            pin.on_edge(self._int_handler, InputPin.FALLING)
        elif _LINUX:
            self._poll_stop = False
            self._poll_thread = _threading.Thread(target=self._poll_loop, daemon=True)
            self._poll_thread.start()

    def off_interrupt(self):
        """Unsubscribe and stop delivery."""
        if self._int_pin_used is not None:
            from periph.connection.input_pin import InputPin
            self._int_pin_used.off_edge(self._int_handler)
            self._int_pin_used = None
        elif _LINUX:
            self._poll_stop = True
        self._callback = None

    def _int_handler(self):
        status = self.poll_interrupt()
        if status[0] and self._callback:
            self._callback(status)

    def _poll_loop(self):
        import time
        was_fault = False
        while not self._poll_stop:
            status = self.poll_interrupt()
            if status[0] and not was_fault and self._callback:
                self._callback(status)
            was_fault = status[0]
            time.sleep(0.005)

    def poll_interrupt(self):
        """Read the fault status — equivalent to read_fault(), does not clear.

        Returns:
            tuple: (fault, ocp, uvlo, ots, ilimit), all bool.
        """
        return self.read_fault()

File: rust/periph/src/chips/motor/drv8830.rs
//! DRV8830 — Low-voltage motor driver with I²C interface (Texas Instruments).
//!
//! H-bridge driver for a single brushed DC motor, controlled entirely over
//! I²C — there are no PWM/direction pins. The host commands a target output
//! *voltage*; the chip PWM-regulates the bridge to hold that average voltage
//! across the winding regardless of supply sag. Nine selectable addresses
//! (`0x60`–`0x68`) via the tri-state `A0`/`A1` strap pins.
//!
//! Voltage conversion follows the datasheet's Table 1:
//! `voltage = VREF * vset / 16` with `VREF` = 1.285 V (typical, ±4%), usable
//! range 0.48 V (`vset` 6) to 5.06 V (`vset` 63) — see
//! `specs/motor/drv8830.md`.
//!
//! ## Interrupts
//!
//! Rust exposes only [`Drv8830Full::poll_interrupt`] (no callback
//! subscription — polling is always caller-managed in this crate's `no_std`
//! Rust drivers). Faults are never cleared implicitly: a latched OCP/ILIMIT
//! fault also disables the H-bridge, so clearing is always an explicit
//! [`Drv8830Full::clear_fault`].

use embedded_hal::i2c::I2c;

const REG_CONTROL: u8 = 0x00;
const REG_FAULT: u8 = 0x01;

const CTRL_IN1: u8 = 0x01;
const CTRL_IN2: u8 = 0x02;

const FAULT_FAULT: u8 = 0x01;
const FAULT_OCP: u8 = 0x02;
const FAULT_UVLO: u8 = 0x04;
const FAULT_OTS: u8 = 0x08;
const FAULT_ILIMIT: u8 = 0x10;
const FAULT_CLEAR: u8 = 0x80;

/// Default 7-bit I²C address (`A0` = `A1` = GND). Valid range `0x60`–`0x68`.
pub const DRV8830_I2C_ADDRESS: u8 = 0x60;
/// Internal reference voltage, typical (datasheet: 1.235–1.335 V).
pub const DRV8830_VREF: f32 = 1.285;
/// Lowest valid `VSET` code (`0x00`–`0x05` are reserved).
pub const DRV8830_VSET_MIN: u8 = 6;
/// Highest `VSET` code (≈5.06 V).
pub const DRV8830_VSET_MAX: u8 = 63;

/// H-bridge state decoded from `IN1`/`IN2`.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Direction {
    /// `IN1`=0, `IN2`=0 — outputs high-Z (standby).
    Coast,
    /// `IN1`=1, `IN2`=0.
    Forward,
    /// `IN1`=0, `IN2`=1.
    Reverse,
    /// `IN1`=1, `IN2`=1 — both outputs high.
    Brake,
}

/// Decoded `CONTROL` register, returned by [`Drv8830Full::read_output`].
#[derive(Clone, Copy, PartialEq, Debug)]
pub struct Output {
    /// Commanded magnitude in V (`0.0` for a reserved `VSET` code).
    pub voltage: f32,
    /// H-bridge state.
    pub direction: Direction,
}

/// Decoded `FAULT` register, returned by [`Drv8830Full::read_fault`].
#[derive(Clone, Copy, PartialEq, Eq, Debug, Default)]
pub struct Fault {
    /// Any fault condition exists.
    pub fault: bool,
    /// Overcurrent (short-circuit) event.
    pub ocp: bool,
    /// Undervoltage lockout.
    pub uvlo: bool,
    /// Overtemperature shutdown.
    pub ots: bool,
    /// Extended current-limit event.
    pub ilimit: bool,
}

/// Errors from [`Drv8830Full::set_output`].
#[derive(Debug)]
pub enum Drv8830Error<E> {
    /// The underlying I²C bus returned an error.
    Bus(E),
    /// `vset` was outside `6`–`63` (codes `0`–`5` are reserved).
    InvalidVset,
}

impl<E> From<E> for Drv8830Error<E> {
    fn from(e: E) -> Self {
        Drv8830Error::Bus(e)
    }
}

/// Map |voltage| to a `VSET` code; `0` means coast (below the `vset`=6 floor).
fn voltage_to_vset(voltage: f32) -> u8 {
    let mag = if voltage < 0.0 { -voltage } else { voltage };
    let vset = (mag * 16.0 / DRV8830_VREF + 0.5) as u32;
    if vset < DRV8830_VSET_MIN as u32 {
        0
    } else if vset > DRV8830_VSET_MAX as u32 {
        DRV8830_VSET_MAX
    } else {
        vset as u8
    }
}

fn vset_to_voltage(vset: u8) -> f32 {
    if vset < DRV8830_VSET_MIN {
        0.0
    } else {
        DRV8830_VREF * vset as f32 / 16.0
    }
}

/// DRV8830 — minimal interface: drive at a regulated voltage, brake, coast.
pub struct Drv8830Minimal<I2C> {
    i2c: I2C,
    addr: u8,
}

impl<I2C: I2c> Drv8830Minimal<I2C> {
    /// Construct the driver and confirm the device answers (one `CONTROL`
    /// read; the chip has no identity register). Makes no register writes —
    /// the POR default already leaves the motor in standby/coast.
    ///
    /// `addr` is `0x60`–`0x68` per the board's `A0`/`A1` strapping.
    pub fn new(mut i2c: I2C, addr: u8) -> Result<Self, I2C::Error> {
        let mut buf = [0u8; 1];
        i2c.write_read(addr, &[REG_CONTROL], &mut buf)?;
        Ok(Self { i2c, addr })
    }

    fn write_reg(&mut self, reg: u8, value: u8) -> Result<(), I2C::Error> {
        self.i2c.write(self.addr, &[reg, value])
    }

    fn read_reg(&mut self, reg: u8) -> Result<u8, I2C::Error> {
        let mut buf = [0u8; 1];
        self.i2c.write_read(self.addr, &[reg], &mut buf)?;
        Ok(buf[0])
    }

    /// Drive the motor at a regulated output voltage.
    ///
    /// `voltage` is signed, in V: positive = forward, negative = reverse,
    /// `0.0` = standby/coast. Writes `VSET` and `IN1`/`IN2` together in one
    /// `CONTROL` write; a magnitude below the ~0.48 V floor coasts, above
    /// ~5.06 V it is clamped.
    pub fn drive(&mut self, voltage: f32) -> Result<(), I2C::Error> {
        let vset = voltage_to_vset(voltage);
        let value = if vset == 0 {
            0x00
        } else if voltage > 0.0 {
            (vset << 2) | CTRL_IN1
        } else {
            (vset << 2) | CTRL_IN2
        };
        self.write_reg(REG_CONTROL, value)
    }

    /// Short-brake the motor (`IN1` = `IN2` = 1, both outputs high).
    pub fn brake(&mut self) -> Result<(), I2C::Error> {
        self.write_reg(REG_CONTROL, CTRL_IN1 | CTRL_IN2)
    }

    /// Put the bridge in standby/coast (`IN1` = `IN2` = 0) — same as `drive(0.0)`.
    pub fn stop(&mut self) -> Result<(), I2C::Error> {
        self.write_reg(REG_CONTROL, 0x00)
    }

    /// Consume the driver and return the underlying I²C bus.
    pub fn release(self) -> I2C {
        self.i2c
    }
}

/// DRV8830 — full interface: extends [`Drv8830Minimal`] with raw `CONTROL`
/// access, output read-back, fault reporting/clearing, and fault polling.
pub struct Drv8830Full<I2C> {
    inner: Drv8830Minimal<I2C>,
}

impl<I2C: I2c> Drv8830Full<I2C> {
    /// Construct the driver; same presence check as [`Drv8830Minimal::new`].
    pub fn new(i2c: I2C, addr: u8) -> Result<Self, I2C::Error> {
        Ok(Self { inner: Drv8830Minimal::new(i2c, addr)? })
    }

    /// Drive at a regulated voltage. Delegates to [`Drv8830Minimal::drive`].
    pub fn drive(&mut self, voltage: f32) -> Result<(), I2C::Error> {
        self.inner.drive(voltage)
    }

    /// Short-brake. Delegates to [`Drv8830Minimal::brake`].
    pub fn brake(&mut self) -> Result<(), I2C::Error> {
        self.inner.brake()
    }

    /// Standby/coast. Delegates to [`Drv8830Minimal::stop`].
    pub fn stop(&mut self) -> Result<(), I2C::Error> {
        self.inner.stop()
    }

    /// Write the `CONTROL` register from raw fields.
    ///
    /// `vset` must be `6`–`63`; reserved codes return
    /// [`Drv8830Error::InvalidVset`] without touching the bus.
    pub fn set_output(&mut self, vset: u8, in1: bool, in2: bool) -> Result<(), Drv8830Error<I2C::Error>> {
        if !(DRV8830_VSET_MIN..=DRV8830_VSET_MAX).contains(&vset) {
            return Err(Drv8830Error::InvalidVset);
        }
        let value = (vset << 2) | if in1 { CTRL_IN1 } else { 0 } | if in2 { CTRL_IN2 } else { 0 };
        self.inner.write_reg(REG_CONTROL, value)?;
        Ok(())
    }

    /// Read back and decode the `CONTROL` register.
    pub fn read_output(&mut self) -> Result<Output, I2C::Error> {
        let ctrl = self.inner.read_reg(REG_CONTROL)?;
        let direction = match ctrl & 0x03 {
            0 => Direction::Coast,
            1 => Direction::Forward,
            2 => Direction::Reverse,
            _ => Direction::Brake,
        };
        Ok(Output { voltage: vset_to_voltage(ctrl >> 2), direction })
    }

    /// Read the `FAULT` register without clearing it.
    pub fn read_fault(&mut self) -> Result<Fault, I2C::Error> {
        let f = self.inner.read_reg(REG_FAULT)?;
        Ok(Fault {
            fault: f & FAULT_FAULT != 0,
            ocp: f & FAULT_OCP != 0,
            uvlo: f & FAULT_UVLO != 0,
            ots: f & FAULT_OTS != 0,
            ilimit: f & FAULT_ILIMIT != 0,
        })
    }

    /// Clear all fault status bits (`CLEAR` = 1); re-enables the H-bridge if
    /// an OCP/ILIMIT fault had latched it off.
    pub fn clear_fault(&mut self) -> Result<(), I2C::Error> {
        self.inner.write_reg(REG_FAULT, FAULT_CLEAR)
    }

    /// Read the fault status — equivalent to [`Self::read_fault`], does not clear.
    pub fn poll_interrupt(&mut self) -> Result<Fault, I2C::Error> {
        self.read_fault()
    }

    /// Consume the driver and return the underlying I²C bus.
    pub fn release(self) -> I2C {
        self.inner.release()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use embedded_hal_mock::eh1::i2c::{Mock as I2cMock, Transaction as I2cTransaction};

    const ADDR: u8 = 0x60;

    #[test]
    fn minimal_drive_brake_stop() {
        let transactions = vec![
            // new(): presence check reads CONTROL.
            I2cTransaction::write_read(ADDR, vec![REG_CONTROL], vec![0x00]),
            // drive(3.0): VSET 37, forward.
            I2cTransaction::write(ADDR, vec![REG_CONTROL, (37 << 2) | 0x01]),
            // drive(-2.0): VSET 25, reverse.
            I2cTransaction::write(ADDR, vec![REG_CONTROL, (25 << 2) | 0x02]),
            // drive(0.4): below the floor -> coast.
            I2cTransaction::write(ADDR, vec![REG_CONTROL, 0x00]),
            // drive(0.48): VSET 6.
            I2cTransaction::write(ADDR, vec![REG_CONTROL, (6 << 2) | 0x01]),
            // drive(9.0): clamped to VSET 63.
            I2cTransaction::write(ADDR, vec![REG_CONTROL, (63 << 2) | 0x01]),
            // brake(), stop().
            I2cTransaction::write(ADDR, vec![REG_CONTROL, 0x03]),
            I2cTransaction::write(ADDR, vec![REG_CONTROL, 0x00]),
        ];
        let mut i2c = I2cMock::new(&transactions);
        let mut motor = Drv8830Minimal::new(i2c.clone(), ADDR).expect("init");
        motor.drive(3.0).expect("drive fwd");
        motor.drive(-2.0).expect("drive rev");
        motor.drive(0.4).expect("drive floor");
        motor.drive(0.48).expect("drive vset6");
        motor.drive(9.0).expect("drive clamp");
        motor.brake().expect("brake");
        motor.stop().expect("stop");
        i2c.done();
    }

    #[test]
    fn full_output_and_faults() {
        let transactions = vec![
            I2cTransaction::write_read(ADDR, vec![REG_CONTROL], vec![0x00]),
            // set_output(20, false, true).
            I2cTransaction::write(ADDR, vec![REG_CONTROL, (20 << 2) | 0x02]),
            // set_output(5, ..) is rejected without a bus transaction.
            // read_output(): VSET 63 forward, VSET 16 reverse, brake.
            I2cTransaction::write_read(ADDR, vec![REG_CONTROL], vec![(63 << 2) | 0x01]),
            I2cTransaction::write_read(ADDR, vec![REG_CONTROL], vec![(16 << 2) | 0x02]),
            I2cTransaction::write_read(ADDR, vec![REG_CONTROL], vec![0x03]),
            // read_fault(): FAULT + ILIMIT; poll_interrupt(): FAULT + OCP + UVLO + OTS.
            I2cTransaction::write_read(ADDR, vec![REG_FAULT], vec![0x11]),
            I2cTransaction::write_read(ADDR, vec![REG_FAULT], vec![0x0F]),
            // clear_fault().
            I2cTransaction::write(ADDR, vec![REG_FAULT, 0x80]),
        ];
        let mut i2c = I2cMock::new(&transactions);
        let mut motor = Drv8830Full::new(i2c.clone(), ADDR).expect("init");

        motor.set_output(20, false, true).expect("set_output");
        assert!(matches!(motor.set_output(5, true, false), Err(Drv8830Error::InvalidVset)));

        let o = motor.read_output().expect("read_output");
        assert_eq!(o.direction, Direction::Forward);
        assert!((o.voltage - 5.06).abs() < 0.01);
        let o = motor.read_output().expect("read_output");
        assert_eq!(o.direction, Direction::Reverse);
        assert!((o.voltage - 1.285).abs() < 0.001);
        let o = motor.read_output().expect("read_output");
        assert_eq!(o, Output { voltage: 0.0, direction: Direction::Brake });

        assert_eq!(
            motor.read_fault().expect("read_fault"),
            Fault { fault: true, ocp: false, uvlo: false, ots: false, ilimit: true }
        );
        assert_eq!(
            motor.poll_interrupt().expect("poll"),
            Fault { fault: true, ocp: true, uvlo: true, ots: true, ilimit: false }
        );
        motor.clear_fault().expect("clear_fault");
        i2c.done();
    }
}
